lib/thrift/clients/binary_framed.ex

```diff
@@ -0,0 +1,240 @@
+    end
+  end
+
+  def disconnect(info, %{sock: sock} = s) do
+    :ok = :gen_tcp.close(sock)
+    case info do
+      {:close, from} ->
+        Connection.reply(from, :ok)
+
+      {:error, :closed} ->
+        Logger.error("Connection closed")
+
+      {:error, reason} ->
+        reason = :inet.format_error(reason)
+        Logger.error("Connection error: #{reason}")
+    end
+    {:connect, :reconnect, %{s | sock: nil}}
+  end
+
+  @spec oneway(pid, data) :: :ok
+  def oneway(conn, data) do
+    Connection.cast(conn, {:oneway, data})
+  end
+
+  @spec request(pid, data, options) :: protocol_response
+  def request(conn, data, options) do
+    socket_opts = Keyword.get(options, :socket_opts, [])
+    gen_server_opts = Keyword.get(options, :gen_server_opts, [])
+    gen_server_timeout = Keyword.get(gen_server_opts, :timeout, 5000)
+
+    Connection.call(conn, {:request, data, socket_opts}, gen_server_timeout)
+  end
+
+  def handle_call(_, _, %{sock: nil} = s) do
+    {:reply, {:error, :closed}, s}
+  end
+
+  def handle_call({:request, data, options}, _, %{sock: sock, timeout: default_timeout} = s) do
+    timeout = Keyword.get(options, :timeout, default_timeout)
+
+    rsp = with :ok <- :gen_tcp.send(sock, data) do
+      :gen_tcp.recv(sock, 0, timeout)
+    end
+
+    case rsp do
+      {:ok, _} = ok ->
+        {:reply, ok, s}
+
+      {:error, :timeout} = timeout ->
+        {:reply, timeout, s}
+
+      {:error, _} = error ->
+        {:disconnect, error, error, s}
+    end
+  end
+
+  def handle_call(:close, from, s) do
+    {:disconnect, {:close, from}, s}
+  end
+
+  def handle_cast(_, %{sock: nil} = s) do
+    {:noreply, s}
+  end
+
+  def handle_cast({:oneway, data}, %{sock: sock} = s) do
+    case :gen_tcp.send(sock, data) do
+      :ok ->
+        {:noreply, s}
+
+      {:error, _} = error ->
+        {:disconnect, error, s}
+    end
+  end
+
+  def deserialize_message_reply(message, rpc_name, sequence_id, reply_module) do
+    Binary.deserialize(:message_begin, message)
+    |> handle_message(sequence_id, rpc_name, reply_module)
+  end
+
+  defp handle_message({:ok, {:reply, sequence_id, rpc_name, decoded_response}},
+                      sequence_id, rpc_name, reply_module) do
+
+    case reply_module.deserialize(decoded_response) do
+      {%{success: nil}=resp, ""} ->
+
+        response = resp
+        |> Map.delete(:__struct__)
+        |> Map.values
+        |> Enum.reject(&is_nil(&1))
+
+        case response do
+          [exception] ->
+            {:error, {:exception, exception}}
+
+          [] ->
+          # This case is when we have a void return on the
+          # remote RPC
+          {:ok, nil}
+        end
+
+      {%{success: success}, ""} when not is_nil(success) ->
+        {:ok, success}
+
+      {resp, extra} ->
+        {:error, {:extraneous_data, resp, extra}}
+    end
+  end
+
+  defp handle_message({:ok, {:exception, sequence_id, rpc_name, response}}, sequence_id, rpc_name, _) do
```

Comment:
no, what's going on here is i'm ensuring the sequence ids and rpc names match. This avoids having a guard clause.
They're "using" themselves.